operator/src/main/java/oracle/kubernetes/operator/JobWatcher.java

```diff
@@ -0,0 +1,232 @@
+// Copyright 2017, 2018, Oracle Corporation and/or its affiliates.  All rights reserved.
+// Licensed under the Universal Permissive License v 1.0 as shown at
+// http://oss.oracle.com/licenses/upl.
+
+package oracle.kubernetes.operator;
+
+import io.kubernetes.client.ApiException;
+import io.kubernetes.client.models.V1Job;
+import io.kubernetes.client.models.V1JobCondition;
+import io.kubernetes.client.models.V1JobStatus;
+import io.kubernetes.client.models.V1ObjectMeta;
+import io.kubernetes.client.util.Watch;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicBoolean;
+import oracle.kubernetes.operator.builders.WatchBuilder;
+import oracle.kubernetes.operator.builders.WatchI;
+import oracle.kubernetes.operator.helpers.CallBuilderFactory;
+import oracle.kubernetes.operator.helpers.ResponseStep;
+import oracle.kubernetes.operator.logging.LoggingFacade;
+import oracle.kubernetes.operator.logging.LoggingFactory;
+import oracle.kubernetes.operator.logging.MessageKeys;
+import oracle.kubernetes.operator.watcher.WatchListener;
+import oracle.kubernetes.operator.work.ContainerResolver;
+import oracle.kubernetes.operator.work.NextAction;
+import oracle.kubernetes.operator.work.Packet;
+import oracle.kubernetes.operator.work.Step;
+
+/** Watches for Jobs to become Ready or leave Ready state */
+public class JobWatcher extends Watcher<V1Job> implements WatchListener<V1Job> {
+  private static final LoggingFacade LOGGER = LoggingFactory.getLogger("Operator", "Operator");
+
+  private final String ns;
+
+  // Map of Pod name to OnReady
+  private final ConcurrentMap<String, OnReady> readyCallbackRegistrations =
```

Comment:
Is OnReady the right concept?  I'd like to understand.  Would think that you would want to know when the Job is complete or failed?